data-prepper-plugins/peer-forwarder/src/main/java/com/amazon/dataprepper/plugins/prepper/peerforwarder/discovery/AwsCloudMapPeerListProvider.java

```diff
@@ -0,0 +1,210 @@
+
+        eventLoop = CommonPools.workerGroup().next();
+        LOG.info("Using AWS CloudMap for Peer Forwarding. namespace='{}', serviceName='{}'",
+                namespaceName, serviceName);
+
+        endpointGroup = new AwsCloudMapDynamicEndpointGroup();
+
+        domainName = serviceName + "." + namespaceName;
+
+        pluginMetrics.gauge(PEER_ENDPOINTS, endpointGroup, group -> group.endpoints().size());
+    }
+
+    static AwsCloudMapPeerListProvider create(PluginSetting pluginSetting, PluginMetrics pluginMetrics) {
+
+        String awsRegion = getRequiredSettingString(pluginSetting, PeerForwarderConfig.AWS_REGION);
+        String namespace = getRequiredSettingString(pluginSetting, PeerForwarderConfig.AWS_CLOUD_MAP_NAMESPACE_NAME);
+        String serviceName = getRequiredSettingString(pluginSetting, PeerForwarderConfig.AWS_CLOUD_MAP_SERVICE_NAME);
+
+
+        Backoff standardBackoff = Backoff.exponential(1000, 32000).withJitter(0.2);
+        int timeToRefreshSeconds = 20;
+
+        ServiceDiscoveryAsyncClient serviceDiscoveryAsyncClient = ServiceDiscoveryAsyncClient
+                .builder()
+                .region(Region.of(awsRegion))
+                .credentialsProvider(DefaultCredentialsProvider.create())
+                .build();
+
+        return new AwsCloudMapPeerListProvider(
+                serviceDiscoveryAsyncClient,
+                namespace,
+                serviceName,
+                timeToRefreshSeconds,
+                standardBackoff,
+                pluginMetrics);
+    }
+
+    private static String getRequiredSettingString(PluginSetting pluginSetting, String propertyName) {
+        String domainName = pluginSetting.getStringOrDefault(propertyName, null);
+        return Objects.requireNonNull(domainName, String.format("Missing '%s' configuration value", propertyName));
+    }
+
+    @Override
+    public List<String> getPeerList() {
+
+        return endpointGroup.endpoints()
+                .stream()
+                .map(Endpoint::ipAddr)
+                .collect(Collectors.toList());
+    }
+
+    @Override
+    public void addListener(Consumer<? super List<Endpoint>> listener) {
+
+        endpointGroup.addListener(listener);
+
+    }
+
+    @Override
+    public void removeListener(Consumer<?> listener) {
+
+        endpointGroup.removeListener(listener);
+
+    }
+
+    @Override
+    public void close() {
+        endpointGroup.close();
+    }
+
+    /**
+     * The {@link DynamicEndpointGroup} class serves as a useful base class for updating
+     * endpoints by supporting the endpoint observer pattern for us. We just need to
+     * periodically check for updates and call {@link DynamicEndpointGroup#setEndpoints(Iterable)}.
+     */
+    private class AwsCloudMapDynamicEndpointGroup extends DynamicEndpointGroup {
+
+        private int failedAttemptCount = 0;
+        private volatile ScheduledFuture<?> scheduledDiscovery;
```

Comment:
Looks like we want this always in the memory ?

Reply:
Yes, this is updated in two different places and we don't know which thread they are running in - 1) After getting the update, we reschedule; 2) Upon closing, we need to cancel the future.